precompiles/staking/method_claim_rewards.go

```diff
@@ -0,0 +1,151 @@
+)
+
+// claimRewards claims all the rewards for a delegator from a validator.
+// As F1 Cosmos distribution scheme implements an all or nothing withdrawal, the precompile will
+// withdraw all the rewards for the delegator, filter ZRC20 and unlock them to the delegator EVM address.
+func (c *Contract) claimRewards(
+	ctx sdk.Context,
+	evm *vm.EVM,
+	_ *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	if len(args) != 2 {
+		return nil, &precompiletypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 2,
+		}
+	}
+
+	delegatorAddr, validatorAddr, err := unpackClaimRewardsArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get delegator Cosmos address.
+	delegatorCosmosAddr, err := precompiletypes.GetCosmosAddress(c.bankKeeper, delegatorAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get validator Cosmos address.
+	validatorCosmosAddr, err := sdk.ValAddressFromBech32(validatorAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Withdraw all the delegation rewards.
+	// The F1 Cosmos distribution scheme implements an all or nothing withdrawal.
+	// The coins could be of multiple denomination, and a mix of ZRC20 and Cosmos coins.
+	coins, err := c.distributionKeeper.WithdrawDelegationRewards(ctx, delegatorCosmosAddr, validatorCosmosAddr)
+	if err != nil {
+		return nil, precompiletypes.ErrUnexpected{
+			When: "WithdrawDelegationRewards",
+			Got:  err.Error(),
+		}
+	}
+
+	// For all the ZRC20 coins withdrawed:
+	// - Check the amount to unlock is valid.
+	// - Burn the Cosmos coins.
+	// - Unlock the ZRC20 coins.
+	for _, coin := range coins {
+		// Filter out invalid coins.
+		if !coin.IsValid() || !coin.Amount.IsPositive() || !precompiletypes.CoinIsZRC20(coin.Denom) {
+			continue
+		}
+
+		// Notice that instead of returning errors we just skip the coin. This is because there might be
+		// more than one ZRC20 coin in the delegation rewards, and we want to unlock as many as possible.
+		// Coins are locked in the bank precompile, so it should be possible to unlock them afterwards.
+		var (
+			zrc20Addr   = common.HexToAddress(strings.TrimPrefix(coin.Denom, config.ZRC20DenomPrefix))
+			zrc20Amount = coin.Amount.BigInt()
+		)
+
+		// Check if bank address has enough ZRC20 balance.
+		// This check is also made inside UnlockZRC20, but repeat it here to avoid burning the coins.
+		if err := c.fungibleKeeper.CheckZRC20Balance(ctx, zrc20Addr, bank.ContractAddress, zrc20Amount); err != nil {
+			ctx.Logger().Info(
+				"Claimed invalid amount of ZRC20 Validator Rewards",
+				"Total", zrc20Amount,
+				"Denom", precompiletypes.ZRC20ToCosmosDenom(zrc20Addr),
+			)
+
+			continue
+		}
+
+		coinSet := sdk.NewCoins(coin)
+
+		// Send the coins to the fungible module to burn them.
+		if err := c.bankKeeper.SendCoinsFromAccountToModule(ctx, delegatorCosmosAddr, fungibletypes.ModuleName, coinSet); err != nil {
+			continue
+		}
+
+		ctx.Logger().Info(
+			"Sentf ZRC20 coins from delegator to Module",
```

Comment:
```suggestion
			"Sent ZRC20 coins from delegator to Module",
```
I think it would make more sense to have just one log for the final action "rewards claimed". Any specific reasoning to have log specifically for this part?

I actually think we could use debug instead for the log in the precompile

Reply:
Agree, makes more sense.